# Nucleotide Encoding of Amino Acid Oligonucleotides
Scripts to encode a designed library of amino acid oligonupeptides as nucleotide oligonucleoties. 

### Dependencies/Requirements
    - If you plan to use the provided DeepLearning model, h2o 3.20.0.8 MUST be installed on your system.
    - OpenMP 
    - g++
    - Python 3.5 or greater
    - Pandas


### Usage
```
Use h2o to select encodings for oligos.

optional arguments:
  -h, --help            show this help message and exit
  -m MODEL, --model MODEL
                        Trained model that ill be used to predict bindings
  -r RATIO_FILE, --ratio_file RATIO_FILE
                        File containing output produced by oligo_encoding
                        script.
  -n NN_SUBSET_SIZE, --nn_subset_size NN_SUBSET_SIZE
                        The nn_subset_size number of encodings will be output
                        that have the smallest nn deviation
  -o OUT_FILE, --out_file OUT_FILE
                        File to write final encodings out to.
  -s SEQUENCES, --sequences SEQUENCES
                        File to read sequences generated by the oligo_encoding
                        script
  -i INPUT, --input INPUT
                        Input file containing name,seq pairs, the maximum line
                        length is 128
  -c CORES, --cores CORES
                        Number of cores to use in the oligo_encoding script,
                        default is 2, as 2 generally has the best performance
  --subsample SUBSAMPLE
                        Number of randomly generated encodings to subsample
  --gc_target GC_TARGET
                        Target gc ratio for generated seqsThe first subsample
                        number of encodings to take will be those with minimum
                        absolute value difference from gc_target.
  -p PROBABILITY_FILE, --probability_file PROBABILITY_FILE
                        probability_file: lines must be formatted as
                        {letter},{nucleotides,3},{weighting},{index}. The
                        weightings do not need to sum to 1. Codon indices must
                        range from 0 to 63.
  -t TRIALS, --trials TRIALS
                        Number of trials to perform. For each sequence in
                        sequences, trials number of candidate encodings will
                        be created.
  --read_per_loop READ_PER_LOOP
                        Number of lines from the output seq and ratio files to
                        read at a time, the higher this parameter is the more
                        memory will be used by h2o.


```	